tree/forest/v7/src/RField.cxx

```diff
@@ -0,0 +1,543 @@
+#include <cctype> // for isspace
+#include <cstdlib> // for malloc, free
+#include <exception>
+#include <iostream>
+#include <utility>
+
+ROOT::Experimental::Detail::RFieldBase::RFieldBase(
+   std::string_view name, std::string_view type, EForestStructure structure, bool isSimple)
+   : fName(name), fType(type), fStructure(structure), fIsSimple(isSimple), fParent(nullptr), fPrincipalColumn(nullptr)
+{
+}
+
+ROOT::Experimental::Detail::RFieldBase::~RFieldBase()
+{
+}
+
+ROOT::Experimental::Detail::RFieldBase*
+ROOT::Experimental::Detail::RFieldBase::Create(const std::string &fieldName, const std::string &typeName)
+{
+   std::string normalizedType(typeName);
+   normalizedType.erase(remove_if(normalizedType.begin(), normalizedType.end(), isspace), normalizedType.end());
+   // TODO(jblomer): use a type translation map
+   if (normalizedType == "Float_t") normalizedType = "float";
+   if (normalizedType == "Double_t") normalizedType = "double";
+   if (normalizedType == "Int_t") normalizedType = "std::int32_t";
+   if (normalizedType == "int") normalizedType = "std::int32_t";
+   if (normalizedType == "unsigned") normalizedType = "std::uint32_t";
+   if (normalizedType == "unsigned int") normalizedType = "std::uint32_t";
+   if (normalizedType == "UInt_t") normalizedType = "std::uint32_t";
+   if (normalizedType == "ULong64_t") normalizedType = "std::uint64_t";
+   if (normalizedType == "string") normalizedType = "std::string";
+   if (normalizedType.substr(0, 7) == "vector<") normalizedType = "std::" + normalizedType;
+
+   if (normalizedType == "ROOT::Experimental::ClusterSize_t") return new RField<ClusterSize_t>(fieldName);
+   if (normalizedType == "std::int32_t") return new RField<std::int32_t>(fieldName);
+   if (normalizedType == "std::uint32_t") return new RField<std::uint32_t>(fieldName);
+   if (normalizedType == "std::uint64_t") return new RField<std::uint64_t>(fieldName);
+   if (normalizedType == "float") return new RField<float>(fieldName);
+   if (normalizedType == "double") return new RField<double>(fieldName);
+   if (normalizedType == "std::string") return new RField<std::string>(fieldName);
+   if (normalizedType.substr(0, 12) == "std::vector<") {
+      std::string itemTypeName = normalizedType.substr(12, normalizedType.length() - 13);
+      auto itemField = Create(GetCollectionName(fieldName), itemTypeName);
+      return new RFieldVector(fieldName, std::unique_ptr<Detail::RFieldBase>(itemField));
+   }
+   // For the time being, we silently read RVec fields as std::vector
+   if (normalizedType.substr(0, 19) == "ROOT::VecOps::RVec<") {
+      std::string itemTypeName = normalizedType.substr(19, normalizedType.length() - 20);
+      auto itemField = Create(GetCollectionName(fieldName), itemTypeName);
+      return new RFieldVector(fieldName, std::unique_ptr<Detail::RFieldBase>(itemField));
+   }
+   // TODO: create an RFieldCollection?
+   if (normalizedType == ":Collection:") return new RField<ClusterSize_t>(fieldName);
```

Comment:
Is it really helpful to pass on the information that the type is a collection via a string?  Wouldn't it be clearer/sturdier to pass is as an explicit parameter?